File: package/umt_go/src/date/format_relative.go
package date

import (
	"fmt"
	"math"
	"strings"
	"time"

	"github.com/riya-amemiya/umt-go/src/consts"
)

// relativeThreshold pairs a relative-time unit name with its length in
// milliseconds, ordered from largest to smallest.
type relativeThreshold struct {
	unit string
	ms   int64
}

var relativeThresholds = []relativeThreshold{
	{"year", consts.OneYearMs},
	{"month", consts.OneMonthMs},
	{"week", consts.OneWeekMs},
	{"day", consts.OneDayMs},
	{"hour", consts.OneHourMs},
	{"minute", consts.OneMinuteMs},
	{"second", consts.OneSecondMs},
}

// FormatRelative formats a date relative to a base date and returns a localized
// relative-time string. It picks the largest unit that fits the absolute delta,
// mirroring Intl.RelativeTimeFormat with numeric "auto".
//
// Only the "en" and "ja" locales are supported; any other locale (including the
// empty string) falls back to "en". The variadic options accept an optional
// base date followed by an optional locale; when the base date is omitted the
// current time is used.
//
// Example:
//
//	FormatRelative(time.Now().Add(-time.Hour), time.Now(), "en") // "1 hour ago"
//	FormatRelative(time.Now().Add(24*time.Hour), time.Now(), "ja") // "明日"
func FormatRelative(date time.Time, baseDate time.Time, locale string) string {
	deltaMs := date.UnixMilli() - baseDate.UnixMilli()
	absDelta := deltaMs
	if absDelta < 0 {
		absDelta = -absDelta
	}

	for _, threshold := range relativeThresholds {
		if absDelta >= threshold.ms {
			value := int(math.Round(float64(deltaMs) / float64(threshold.ms)))
			return formatRelativeUnit(value, threshold.unit, locale)
		}
	}
	return formatRelativeUnit(0, "second", locale)
}

// formatRelativeUnit renders a single value/unit pair using numeric "auto"
// semantics, where -1, 0, and +1 of certain units produce idiomatic words.
func formatRelativeUnit(value int, unit, locale string) string {
	if strings.HasPrefix(locale, "ja") {
		return formatRelativeJa(value, unit)
	}
	return formatRelativeEn(value, unit)
}

// formatRelativeEn renders English relative-time strings matching CLDR's
// "auto" numbering, including special words such as "now", "tomorrow", and
// "last week".
func formatRelativeEn(value int, unit string) string {
	if word, ok := relativeAutoEn[unitKey{unit, value}]; ok {
		return word
	}

	plural := unit
	if value != 1 && value != -1 {
		plural = unit + "s"
	}
	if value < 0 {
		return fmt.Sprintf("%d %s ago", -value, plural)
	}
	return fmt.Sprintf("in %d %s", value, plural)
}

// formatRelativeJa renders Japanese relative-time strings matching CLDR's
// "auto" numbering, including special words such as "今" and "明日".
func formatRelativeJa(value int, unit string) string {
	if word, ok := relativeAutoJa[unitKey{unit, value}]; ok {
		return word
	}

	counter := relativeUnitJa[unit]
	if value < 0 {
		return fmt.Sprintf("%d %s前", -value, counter)
	}
	return fmt.Sprintf("%d %s後", value, counter)
}

// unitKey identifies a special-cased value/unit combination.
type unitKey struct {
	unit  string
	value int
}

// relativeAutoEn holds the idiomatic English words used by numeric "auto".
var relativeAutoEn = map[unitKey]string{
	{"second", 0}: "now",
	{"minute", 0}: "this minute",
	{"hour", 0}:   "this hour",
	{"day", -1}:   "yesterday",
	{"day", 0}:    "today",
	{"day", 1}:    "tomorrow",
	{"week", -1}:  "last week",
	{"week", 0}:   "this week",
	{"week", 1}:   "next week",
	{"month", -1}: "last month",
	{"month", 0}:  "this month",
	{"month", 1}:  "next month",
	{"year", -1}:  "last year",
	{"year", 0}:   "this year",
	{"year", 1}:   "next year",
}

// relativeAutoJa holds the idiomatic Japanese words used by numeric "auto".
var relativeAutoJa = map[unitKey]string{
	{"second", 0}: "今",
	{"minute", 0}: "1 分以内",
	{"hour", 0}:   "1 時間以内",
	{"day", -2}:   "一昨日",
	{"day", -1}:   "昨日",
	{"day", 0}:    "今日",
	{"day", 1}:    "明日",
	{"day", 2}:    "明後日",
	{"week", -1}:  "先週",
	{"week", 0}:   "今週",
	{"week", 1}:   "来週",
	{"month", -1}: "先月",
	{"month", 0}:  "今月",
	{"month", 1}:  "来月",
	{"year", -1}:  "昨年",
	{"year", 0}:   "今年",
	{"year", 1}:   "来年",
}

// relativeUnitJa maps a unit name to its Japanese counter for non-special
// values.
var relativeUnitJa = map[string]string{
	"second": "秒",
	"minute": "分",
	"hour":   "時間",
	"day":    "日",
	"week":   "週間",
	"month":  "か月",
	"year":   "年",
}

File: package/umt_rust/src/date/format_relative.rs
//! Relative-time formatting utility.

use chrono::{DateTime, Utc};

use crate::consts::clock::{
    ONE_DAY_MS, ONE_HOUR_MS, ONE_MINUTE_MS, ONE_MONTH_MS, ONE_SECOND_MS, ONE_WEEK_MS, ONE_YEAR_MS,
};

/// A relative-time unit, mirroring the subset of `Intl.RelativeTimeFormat`
/// units used by the threshold table.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum RelativeUnit {
    Year,
    Month,
    Week,
    Day,
    Hour,
    Minute,
    Second,
}

const THRESHOLDS: [(RelativeUnit, u64); 7] = [
    (RelativeUnit::Year, ONE_YEAR_MS),
    (RelativeUnit::Month, ONE_MONTH_MS),
    (RelativeUnit::Week, ONE_WEEK_MS),
    (RelativeUnit::Day, ONE_DAY_MS),
    (RelativeUnit::Hour, ONE_HOUR_MS),
    (RelativeUnit::Minute, ONE_MINUTE_MS),
    (RelativeUnit::Second, ONE_SECOND_MS),
];

/// Rounds a floating-point value the same way as JavaScript `Math.round`
/// (half rounds toward positive infinity).
fn js_round(value: f64) -> i64 {
    (value + 0.5).floor() as i64
}

fn unit_singular(unit: RelativeUnit) -> &'static str {
    match unit {
        RelativeUnit::Year => "year",
        RelativeUnit::Month => "month",
        RelativeUnit::Week => "week",
        RelativeUnit::Day => "day",
        RelativeUnit::Hour => "hour",
        RelativeUnit::Minute => "minute",
        RelativeUnit::Second => "second",
    }
}

/// Special wording for `numeric: "auto"` at value +1 / -1, returning `None`
/// when the unit has no special phrasing (hour, minute, second).
fn auto_special(unit: RelativeUnit, value: i64) -> Option<&'static str> {
    match (unit, value) {
        (RelativeUnit::Year, 1) => Some("next year"),
        (RelativeUnit::Year, -1) => Some("last year"),
        (RelativeUnit::Month, 1) => Some("next month"),
        (RelativeUnit::Month, -1) => Some("last month"),
        (RelativeUnit::Week, 1) => Some("next week"),
        (RelativeUnit::Week, -1) => Some("last week"),
        (RelativeUnit::Day, 1) => Some("tomorrow"),
        (RelativeUnit::Day, -1) => Some("yesterday"),
        _ => None,
    }
}

/// Formats a `value`/`unit` pair the way `Intl.RelativeTimeFormat("en",
/// { numeric: "auto" })` would.
fn format_en_auto(value: i64, unit: RelativeUnit) -> String {
    if value == 0 && unit == RelativeUnit::Second {
        return "now".to_string();
    }
    if let Some(special) = auto_special(unit, value) {
        return special.to_string();
    }

    let abs = value.abs();
    let mut label = unit_singular(unit).to_string();
    if abs != 1 {
        label.push('s');
    }

    if value > 0 {
        format!("in {} {}", abs, label)
    } else {
        format!("{} {} ago", abs, label)
    }
}

/// Formats a date relative to a base date.
///
/// Picks the largest unit that fits the absolute delta and formats it the
/// way `Intl.RelativeTimeFormat(locale, { numeric: "auto" })` would for the
/// English (`"en"`) locale. Locales other than `"en"` fall back to English
/// wording, because the Rust standard library and `chrono` do not provide
/// `Intl`-style locale data.
///
/// # Arguments
///
/// * `date` - The target date
/// * `base_date` - Reference date
/// * `_locale` - BCP 47 locale tag (only `"en"` is fully supported)
///
/// # Returns
///
/// A relative-time string.
///
/// # Examples
///
/// ```
/// use chrono::{TimeZone, Utc};
/// use umt_rust::date::umt_format_relative;
///
/// let base = Utc.with_ymd_and_hms(2025, 4, 15, 12, 0, 0).unwrap();
/// let target = base - chrono::Duration::seconds(5);
/// assert_eq!(umt_format_relative(&target, &base, Some("en")), "5 seconds ago");
///
/// let target = base + chrono::Duration::days(1);
/// assert_eq!(umt_format_relative(&target, &base, Some("en")), "tomorrow");
/// ```
pub fn umt_format_relative(
    date: &DateTime<Utc>,
    base_date: &DateTime<Utc>,
    _locale: Option<&str>,
) -> String {
    let delta_ms = date.timestamp_millis() - base_date.timestamp_millis();
    let abs_delta = delta_ms.unsigned_abs();

    for (unit, ms) in THRESHOLDS {
        if abs_delta >= ms {
            let value = js_round(delta_ms as f64 / ms as f64);
            return format_en_auto(value, unit);
        }
    }
    format_en_auto(0, RelativeUnit::Second)
}
